front/FrontService.cpp

```diff
@@ -19,5 +19,478 @@
+        std::make_shared<Error>(CommonError::TIMEOUT, "front service stopped");
+
+    {
+      RecursiveGuard l(x_callback);
+
+      for (auto &callback : m_callback) {
+        FRONT_LOG(INFO) << LOG_DESC("front service stopped")
+                        << LOG_KV("message uuid", callback.first);
+        // cancel the timer first
+        callback.second->timeoutHandler->cancel();
+        callback.second->callbackFunc(errorPtr, nullptr, bytesConstRef(),
+                                      std::function<void(bytesConstRef)>());
+      }
+    }
+
+  } catch (const std::exception &e) {
+    std::cerr << e.what() << '\n';
+  }
+
+  FRONT_LOG(INFO) << LOG_BADGE("stop end") << LOG_KV("nodeID", m_nodeID->hex())
+                  << LOG_KV("groupID", m_groupID);
+
+  return;
+}
+
+/**
+ * @brief: get nodeID list
+ * @return void
+ */
+void FrontService::asyncGetNodeIDs(
+    std::function<void(Error::Ptr _error,
+                       std::shared_ptr<const crypto::NodeIDs> _nodeIDs)>
+        _callback) const {
+  auto errorPtr = std::make_shared<Error>(CommonError::SUCCESS, "success");
+  {
+    RecursiveGuard l(x_nodeIDs);
+    _callback(errorPtr, m_nodeIDs);
+  }
+}
+
+/**
+ * @brief: send message to node
+ * @param _moduleID: moduleID
+ * @param _nodeID: the receiver nodeID
+ * @param _data: send data
+ * @param _timeout: the timeout value of async function, in milliseconds.
+ * @param _callbackFunc: _callbackFunc
+ * @return void
+ */
+void FrontService::asyncSendMessageByNodeID(int _moduleID,
+                                            bcos::crypto::NodeIDPtr _nodeID,
+                                            bytesConstRef _data,
+                                            uint32_t _timeout,
+                                            CallbackFunc _callbackFunc) {
+
+  try {
+    std::string uuid;
+
+    if (_callbackFunc) {
+      uuid = boost::uuids::to_string(boost::uuids::random_generator()());
+
+      auto callback = std::make_shared<Callback>();
+      callback->callbackFunc = _callbackFunc;
+
+      if (_timeout > 0) {
+        // create new timer to handle timeout
+        auto timeoutHandler = std::make_shared<boost::asio::deadline_timer>(
+            *m_ioService, boost::posix_time::milliseconds(_timeout));
+
+        callback->timeoutHandler = timeoutHandler;
+        // callback->startTime = utcSteadyTime();
+        timeoutHandler->async_wait([frontServicePtr = shared_from_this(),
+                                    uuid](const boost::system::error_code &e) {
+          frontServicePtr->onMessageTimeout(e, uuid);
+        });
+      }
+
+      addCallback(uuid, callback);
+
+      FRONT_LOG(DEBUG) << LOG_BADGE("asyncSendMessageByNodeID")
+                       << LOG_KV("uuid", uuid) << LOG_KV("groupID", m_groupID)
+                       << LOG_KV("nodeID", m_nodeID->hex())
+                       << LOG_KV("timeout", _timeout);
+
+    } // if (_callback)
+
+    onSendMessage(_moduleID, _nodeID, uuid, _data);
```

Comment:
uuid is not initialized

Reply:
if (_callbackFunc) {
      uuid = boost::uuids::to_string(boost::uuids::random_generator()());

    uuid will be set only _callbackFunc is present